Java/core/src/main/java/com/amazon/randomcutforest/PredictiveRandomCutForest.java

```diff
@@ -0,0 +1,377 @@
+        preprocessorBuilder.dimensions(dimensions);
+        preprocessorBuilder.stopNormalization(builder.stopNormalization.orElse(DEFAULT_STOP_NORMALIZATION));
+        preprocessorBuilder.startNormalization(builder.startNormalization.orElse(DEFAULT_START_NORMALIZATION));
+
+        preprocessor = preprocessorBuilder.build();
+    }
+
+    public PredictiveRandomCutForest(ForestMode forestMode, TransformMethod method, IPreprocessor preprocessor,
+            RandomCutForest forest) {
+        this.forestMode = forestMode;
+        this.transformMethod = method;
+        this.preprocessor = preprocessor;
+        this.forest = forest;
+    }
+
+    void validateNonNegativeArray(double[] array) {
+        if (array != null) {
+            for (double element : array) {
+                checkArgument(element >= 0, " has to be non-negative");
+            }
+        }
+    }
+
+    public SampleSummary predict(float[] inputPoint, long timestamp, int[] missingValues) {
+        return predict(inputPoint, timestamp, missingValues, 5, 0.3, 0.7, false);
+    }
+
+    public SampleSummary predict(float[] inputPoint, long timestamp, int[] missingValues, int numberOfRepresentatives,
+            double shrinkage, double centrality, boolean project) {
+
+        SampleSummary summary = null;
+        boolean cacheDisabled = (forest.getBoundingBoxCacheFraction() == 0);
+        try {
+            if (cacheDisabled) { // turn caching on temporarily
+                forest.setBoundingBoxCacheFraction(1.0);
+            }
+            if (missingValues != null) {
+                checkArgument(missingValues.length <= inputPoint.length, " incorrect data");
+                for (int i = 0; i < missingValues.length; i++) {
+                    checkArgument(missingValues[i] >= 0, " missing values cannot be at negative position");
+                    checkArgument(missingValues[i] < inputPoint.length,
+                            "missing values cannot be at position larger than input length");
+                }
+            }
+            float[] scaledInput = preprocessor.getScaledInput(inputPoint, timestamp);
+            summary = preprocessor.invertSummary(
+                    forest.getConditionalFieldSummary(scaledInput, missingValues.length, missingValues,
+                            numberOfRepresentatives, shrinkage, true, project, centrality),
+                    missingValues.length, missingValues, scaledInput);
+        } finally {
+            if (cacheDisabled) { // turn caching off
+                forest.setBoundingBoxCacheFraction(0);
+            }
+        }
+        return summary;
+    }
+
+    public void update(float[] record, long timestamp) {
+        float[] scaled = preprocessor.getScaledInput(record, timestamp);
+        preprocessor.update(toDoubleArray(record), scaled, timestamp, null, forest);
+        if (scaled != null) {
+            forest.update(scaled);
```

Comment:
No - the preprocessor.update would already address the update. We do not need to repeat the update (it'll effectively approximately halve the sample size). Fixed. ImputeProcessor updates the forest via generateShingle() -- because it needs to regenerate the shingle (which itself may have missing values).